{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ae06e520",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-11T21:59:57.142781Z",
     "start_time": "2023-01-11T21:59:57.137018Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "F=\"shuf-mutations-uniq-woRMSK.GENCODE-exon_no-overlap-ref.only-9-EXPANDED-500.bed\"\n",
    "PATH = \"/wynton/home/ahituv/fongsl/nullomers/data/shuffle\"\n",
    "TEST_BED =os.path.join(PATH, F)\n",
    "BUILD = \"hg38\"\n",
    "OUTDIR = \"/wynton/home/ahituv/fongsl/tools/unit_test/out\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5e19edad",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-11T22:01:27.692416Z",
     "start_time": "2023-01-11T22:01:27.682801Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nFrom bed file w DNA sequence str, calculate GC count, GC %, GC dinucleotide content\\n\\nInput: \\n    .bed file (str) - full path to bed file\\n    -b (str) - genome build. Default is hg38\\n    -o (str) - full path to output directory\\n\\nMethod \\n    1. instantiate and parse arguments\\n    2. \\n\\n\\nOutput: \\n    .tsv with sequence id,  GC count, and GC density\\n\\nNotes\\n- turns .bed -> .fa, then evaluates GC\\n- human DNA only\\n- writes .fa file to same directory as .bed file\\n'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import argparse\n",
    "\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "\n",
    "import gc_content as gc\n",
    "import glob\n",
    "from joblib import Parallel, delayed\n",
    "import os, sys\n",
    "\n",
    "import subprocess as sp\n",
    "\n",
    "\"\"\"\n",
    "From bed file w DNA sequence str, calculate GC count, GC %, GC dinucleotide content\n",
    "\n",
    "Input: \n",
    "    .bed file (str) - full path to bed file\n",
    "    -b (str) - genome build. Default is hg38\n",
    "    -o (str) - full path to output directory\n",
    "\n",
    "Method \n",
    "    1. instantiate and parse arguments\n",
    "    2. \n",
    "\n",
    "\n",
    "Output: \n",
    "    .tsv with sequence id,  GC count, and GC density\n",
    "\n",
    "Notes\n",
    "- turns .bed -> .fa, then evaluates GC\n",
    "- human DNA only\n",
    "- writes .fa file to same directory as .bed file\n",
    "\"\"\"\n",
    "\n",
    "###\n",
    "#   arguments\n",
    "###\n",
    "\n",
    "arg_parser = argparse.ArgumentParser(description=\"turn .bed into .fa, chromosome by chromosome, get GC content\")\n",
    "\n",
    "arg_parser.add_argument(\"bed\", type=str, help='.bed file w/ full path')\n",
    "arg_parser.add_argument(\"-b\", \"--build\", type=str, default='hg38', choices=['hg19', 'hg38'],\n",
    "                        help='human genome assembly; default=hg38')\n",
    "arg_parser.add_argument(\"-o\", \"--outdirectory\", type=str, help='directory to write gc information')\n",
    "\n",
    "# parse the arguments \n",
    "args = arg_parser.parse_args()\n",
    "\n",
    "TEST_BED = args.bed\n",
    "BUILD = args.build\n",
    "OUTDIR = args.outdirectory\n",
    "\n",
    "### \n",
    "# Functions\n",
    "### \n",
    "\n",
    "def splitFile(test_file):\n",
    "    \"\"\"\n",
    "    split test file into path, filename, and sample_id\n",
    "    \n",
    "    requirements\n",
    "        os \n",
    "        \n",
    "    input\n",
    "        test_file (str) - full path + test_file name\n",
    "\n",
    "    method\n",
    "        1. path, file = os.splitpath\n",
    "        2. filename = os.splittext\n",
    "    \n",
    "    return \n",
    "        path, file, filename (strs)\n",
    "    \"\"\"\n",
    "    \n",
    "    path, file = os.path.split(test_file)\n",
    "    \n",
    "    filename = os.path.splitext(file)[0]\n",
    "\n",
    "    return path, file, filename\n",
    "    \n",
    "    \n",
    "def splitByChr(test_bed, filename, outdir):\n",
    "        \"\"\"\n",
    "        Split a file by chromosome name in awk \n",
    "        \n",
    "        input\n",
    "            test_bed (str) - .bed file w/ full path to split by chromosome\n",
    "            outdir (str) - path to write output files\n",
    "            \n",
    "        method \n",
    "    \n",
    "            1. change directory to outdirectory\n",
    "            2. awk split on chromosome column\n",
    "            3. glob resultant files.\n",
    "        \n",
    "        return \n",
    "            bed_chr_list (list) - list of chr-split .bed files\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        if os.getcwd()!=outdir:\n",
    "            #1 go to bed dir\n",
    "            os.chdir(outdir)\n",
    "        \n",
    "        #2 split test into chrN.bed files\n",
    "        \n",
    "    \n",
    "        # old awk\n",
    "        cmd = \"awk '{print >$1\\\"_%s.bed\\\"}' %s\" % (filename, test_bed) \n",
    "        print(cmd)\n",
    "        os.system(cmd)\n",
    "        #args = [\"awk\", '{print >$1\\\"_%s_temp.bed\\\"}' % filename, test_bed]\n",
    "        #p = sp.Popen(args, stdin = sp.PIPE, stdout = sp.PIPE, stderr = sp.PIPE)\n",
    "        #print(p.stdout.readline()) # read the first line\n",
    "\n",
    "        #3 glob chr split bed files. \n",
    "        bed_chr_list = glob.glob(os.path.join(outdir, f\"chr*_{filename}.bed\"))  # glob chromosomes\n",
    "\n",
    "        return bed_chr_list\n",
    "    \n",
    "    \n",
    "def getDNA(genome_build):\n",
    "    \n",
    "    \"\"\"\n",
    "    return path to chr.fa.gz files\n",
    "    \"\"\"\n",
    "    \n",
    "    if \"capra\" in os.getcwd():    \n",
    "        dna_path = f\"/dors/capra_lab/data/dna/human/{genome_build}\"\n",
    "\n",
    "    else:\n",
    "        dna_path = f\"/wynton/home/ahituv/fongsl/dna/{genome_build}\"\n",
    "    \n",
    "    return dna_path\n",
    "\n",
    "\n",
    "def getFasta(bed_chr, dna_path, outdir):\n",
    "    \"\"\"\n",
    "    get fasta from bed per chromosome \n",
    "    \n",
    "    requirement\n",
    "        bedtools getfasta\n",
    "    \n",
    "    input\n",
    "        bed_chr (str) - chr.bed file w full path\n",
    "        dna_path (str) - path to chr.fa files\n",
    "        outdir (str) - path to write .fa outfile to\n",
    "\n",
    "    method\n",
    "        1. str split for filename, chr_num, \n",
    "        2. instantiate fasta input and output files, zipped input file\n",
    "        3. unzip source chr.fa file if zipped. \n",
    "        4. use bedtools getfasta to get fasta from .bed\n",
    "        5. clean up the bed file\n",
    "        \n",
    "    \"\"\"\n",
    "    #1\n",
    "    filename = (bed_chr.split(\"/\")[-1]).split(\".bed\")[0]\n",
    "    chr_num = filename.split(\"_\")[0]\n",
    "   \n",
    "    #2\n",
    "    fasta_in = os.path.join(dna_path, f\"{chr_num}.fa\") \n",
    "    fasta_out = os.path.join(outdir, f\"{filename}.fa\")  # write\n",
    "    \n",
    "    zipped_in = fasta_in + \".gz\"\n",
    "   \n",
    "    #3\n",
    "    if os.path.exists(zipped_in) is True:\n",
    "        print(\"unzipping\", zipped_in)\n",
    "        sp.call(f\"gunzip {zipped_in}\", shell=True)\n",
    "\n",
    "    \n",
    "    #4 \n",
    "    cmd = f\"bedtools getfasta -fi {fasta_in} -bed {bed_chr} -fo {fasta_out}\"\n",
    "    \n",
    "    if os.path.exists(fasta_out) is False:\n",
    "        sp.call(cmd, shell=True)\n",
    "    \n",
    "        line_count = sum(1 for line in open(fasta_out, \"r\").readlines())\n",
    "\n",
    "        #5 clean up chr temp file\n",
    "        if line_count > 0:\n",
    "            os.remove(bed_chr) \n",
    "\n",
    "            \n",
    "def combineChrs(outdir, filename):\n",
    "    \"\"\"\n",
    "    cat chr.fa files and delete individual files\n",
    "    \"\"\"\n",
    "\n",
    "    # file to write\n",
    "    out_fasta = os.path.join(outdir, f\"{filename}.fa\")\n",
    "\n",
    "    # concat all the chromosomes\n",
    "    cat_handle = os.path.join(outdir, f\"chr*{filename}*.fa\")\n",
    "    \n",
    "    os.system(f\"cat {cat_handle} > {out_fasta}\")\n",
    "\n",
    "    # remove chromosome fasta files. \n",
    "    os.system(f\"rm {cat_handle}\")\n",
    "    \n",
    "    return out_fasta\n",
    "\n",
    "### \n",
    "# GC Functions\n",
    "### \n",
    "\n",
    "\n",
    "def countGC(sequence):\n",
    "    \"\"\"\n",
    "    return gc_fraction, gc_dinucleotide content fraction (custom script) \n",
    "    \"\"\"\n",
    "\n",
    "    # dinucleotide info\n",
    "    dinuc_count, dinuc_frac = gc.count_dinucleotide(sequence)\n",
    "    \n",
    "    return gc_fraction(sequence), dinuc_frac\n",
    "\n",
    "\n",
    "def writeRow(outf, rows):\n",
    "    \n",
    "    \"\"\"\n",
    "    write to outfile list of rows, one row per index. \n",
    "    \"\"\"\n",
    "    \n",
    "    with open(outf, \"w\") as writer:\n",
    "        for row in rows:\n",
    "            writer.write(row)\n",
    "        writer.close()\n",
    "        \n",
    "def gcFromFasta(fasta_file, filename, outdir):\n",
    "    \"\"\"\n",
    "    format the fasta file into GC content\n",
    "    \n",
    "    require \n",
    "        Bio.SeqIO.FastaIO.SimpleFastaParser\n",
    "        \n",
    "    input\n",
    "        fasta_file (str) - path to fasta \n",
    "        filename (str)-name of file\n",
    "        outdir (str) - path to write to \n",
    "        \n",
    "    method \n",
    "        1. make output file\n",
    "        2. instantiate rows, already_processed lists\n",
    "        3. parse through fasta w/ SimpleFastaParser\n",
    "            get seqid, sequence\n",
    "        4. calculate gc content (fraction G|C, fraction G+C)\n",
    "        5. append data to lists\n",
    "        6. write gc content information to file\n",
    "\n",
    "    return \n",
    "        outfile (str) - written gc content file. \n",
    "    \"\"\"\n",
    "    \n",
    "    #1\n",
    "    outf = os.path.join(outdir, f\"{filename}_GC.tsv\" ) # path to write\n",
    "    print(\"writing GC content\", outf)\n",
    "    \n",
    "    #2\n",
    "    rows, already_processed = [], []  # collect the sequence ids that have already been processed. \n",
    " \n",
    "    #3\n",
    "    with open(fasta_file, \"r\") as ff:\n",
    "        for values in SimpleFastaParser(ff):\n",
    "            seq_id, sequence = values \n",
    "            \n",
    "            #4\n",
    "            gc_frac, gc_dinuc_frac = countGC(sequence.upper()) # count GC content, density for sequence\n",
    "\n",
    "            row = f\"{seq_id}\\t{gc_frac}\\t{gc_dinuc_frac}\\n\"\n",
    "                \n",
    "            if seq_id not in already_processed:\n",
    "                    \n",
    "                #5 append row to rows, seq id to already_processed lists\n",
    "                rows.append(row), already_processed.append(seq_id) \n",
    "                        \n",
    "            else:\n",
    "                print(f\"\\nCHECK .FA REDUNDANCY. Seq_id {seq_id} is already written.\\n\")\n",
    "                    \n",
    "    #6\n",
    "    writeRow(outf, rows)\n",
    "                   \n",
    "    return outf\n",
    "    \n",
    "\n",
    "def main(argv):\n",
    "    \n",
    "\n",
    "    # str split file name\n",
    "    path, file, filename = splitFile(TEST_BED)\n",
    "    \n",
    "    # path to input DNA fasta files \n",
    "    dna_path = getDNA(BUILD)\n",
    "    \n",
    "    # split bed file by chromosome\n",
    "    bed_chr_list = splitByChr(TEST_BED, filename, OUTDIR)\n",
    "\n",
    "    # run get fasta per chromosome. \n",
    "    num_cores = len(bed_chr_list)\n",
    "\n",
    "    # turn bed into fasta using bedtools getfasta command\n",
    "    Parallel(n_jobs=num_cores, verbose=100, prefer=\"threads\")(delayed(getFasta)(i, dna_path, OUTDIR) for i in bed_chr_list)\n",
    "\n",
    "    # combine all chr fastas together\n",
    "    FASTA = combineChrs(OUTDIR, filename)\n",
    "    \n",
    "\n",
    "    # get GC fraction per bed file\n",
    "    out_fasta = gcFromFasta(FASTA, filename, OUTDIR)\n",
    "    \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2f12291f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-13T22:30:47.769912Z",
     "start_time": "2022-04-13T22:30:47.757452Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/dors/capra_lab/users/fongsl/tools/unit_test', 'test.bed', 'test')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "path, filename= os.path.split(TEST_BED)\n",
    "split = os.path.splitext(filename)[0]\n",
    "path, filename, split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6424287e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (wynenv)",
   "language": "python",
   "name": "wynenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
